### ISSUE TITLE
File conflicts: Symlinked directories -> same file replaced by real directories -> unique files

### ISSUE THREAD

**Issue by markdirish** [rpm-software-management/rpm#1458]
RPM Version: 4.13.1 (also recreated in 4.15.1)

Hi, my name is Mark Irish and I am on a team of developers that creates and distributes RPMs for open-source software on IBM i. In our distribution of RPM itself, we are trying to replace generic platform macros shared between all versions/releases of our system with release-specific macro files.

### The issue we are running into:

Previously, our `rpm.spec` handled platform macros by creating a symlink from every version/release to a common directory that stored the `macros` file. This meant that `.../rpm/platform/ppc-ibmi7.1/`, `.../rpm/platform/ppc-ibmi7.2/`, and `.../rpm/platform/ppc-ibmi7.3/` all were symlinked to `.../rpm/platform/ppc-ibmi`, and this latter directory held the `macros` file. That worked well, but now we want to enable release-specific flags in our macros.

However, when trying to replace the 3 macro "files" that were all the same (because they were symlinked to the same directory) with 3 different macro files in non-symlinked directories, the 2nd and 3rd file throw a `conflicts between attempted installs` error:

```
Transaction Check Error:
  file /QOpenSys/pkgs/lib/rpm-test/platform/ppc-ibmi7.2/macros conflicts between attempted installs of rpm-test-0.0.0-1.ppc64 and rpm-test-0.0.0-1.ppc64
  file /QOpenSys/pkgs/lib/rpm-test/platform/ppc-ibmi7.3/macros conflicts between attempted installs of rpm-test-0.0.0-1.ppc64 and rpm-test-0.0.0-1.ppc64
```

I don't believe this is an issue with symlinked directories. **We are aware that there is some lua magic that has to be done to convert symlinks to directories.** [We have followed the script here](https://fedoraproject.org/wiki/Packaging:Directory_Replacement#:~:text=Due%20to%20a%20known%20limitation,trying%20to%20install%20the%20package.), and it has fixed previous errors that we were encountering.

The issue seems to be RPM expecting any change to `.../rpm/platform/ppc-ibmi7.1/macros`, `.../rpm/platform/ppc-ibmi7.2/macros`, and `.../rpm/platform/ppc-ibmi7.3/macros` to be uniform: It doesn't complain about `7.1`, the first directory that changes to a new file, only about `7.2` and `7.3`. This also explains why the error notes of `conflicts between attempted installs of` the same version.

**I am able to provide 2 minimal spec files that recreate this issue.**

### The question:

Is this a bug in RPM? It seems possible that this is a strange edge case that wasn't considered, though the area of the code notes https://github.com/rpm-software-management/rpm/blob/4cbdd7c94071f52dfe64466df80e859193d89d89/lib/transaction.c#L628

Is there any RPM or lua magic that can be done to resolve these errors, similar to the lua script to resolve? Possibly create an intermediary release that somehow fixes the issue and allows the next release to replace the single `macros` file with multiple unique files?

I think a solution like `https://github.com/rpm-software-management/rpm/issues/1118` would work, but again that upgrade would have to be an intermediary step, as our issue is errors using RPM to upgrade RPM.

Thank you.

**Comment by Conan-Kudo** [comment]
This is a long-standing issue with file<->directory replacement. There's some guidelines on how to work around the issue here: https://docs.fedoraproject.org/en-US/packaging-guidelines/Directory_Replacement/

**Comment by ffesti** [comment]
Ok, I really thought we already had an upstream ticket for this...

Anyway here my thought on what needs to be done to make this actually work:

Symlink to directory
--------------------
 * New files
   * Collision check needs to take new location into account
 * To be removed files symlink->/dir/file
   * Rewrite location to physical one
   * Change action to remove even on collision with new files
 * Remaining files no longer accessible
   * Move to new dir?
   * Error out?
   * Change to physical location in RPMDB?

Directory to symlink to existing/new dir
------------------------------------
May need to ensure the target dir gets created early
 * New files
   * Collision check needs to take new location into account
 * To be removed files
   * Rename dir	
   *  Rewrite location to new physical one
   * Change action to remove even on collision with new files
 * Remaining, now homeless files
   * Move to new dir if it exist (this is kinda iffy)

Issues
------

* Moving files around
  * Requires redoing the free disk space check
  * May create issues with aliasing - especially	if file	is used	from location unknown to rpm.
Probably more...

**Comment by dmnks** [comment]
Yes, this issue is still present on master, the associated logic hasn't been changed.

**Comment by dmnks** [comment]
This indeed looks like a dupe of https://bugzilla.redhat.com/show_bug.cgi?id=1936422 which #1684 was attempting to fix. The actual, real transaction (assuming that a `%pretrans` scriptlet removing the old symlink is present) does the right thing, this is just about the *test* transaction (which DNF/YUM perform) where it fails.

A minimal package reproducing the issue:
```
%bcond_with new

Name: symlink
%if %{with new}
Version: 2
%else
Version: 1
%endif
Release: 1%{?dist}
Summary: Symlink conflict test
License: GPL

%description
%{summary}.

%if %{with new}
%pretrans -p <lua>
path = "/opt/symlink/bar"
st = posix.stat(path)
if st and st.type == "link" then
  os.remove(path)
end
%endif

%prep
%build
%install

mkdir -p %{buildroot}/opt/symlink/foo
echo hello > %{buildroot}/opt/symlink/foo/blah

%if %{with new}
mkdir -p %{buildroot}/opt/symlink/bar
echo new > %{buildroot}/opt/symlink/bar/blah
%else
ln -s foo %{buildroot}/opt/symlink/bar
%endif


%files
/opt/symlink
```

Usage:
```
$ rpmbuild -bb /path/to/symlink.spec # builds symlink-1-1 package
$ rpmbuild -bb --with new /path/to/symlink.spec # builds symlink-2-1 package
$ rpm -Uhv /path/to/symlink-1-1.rpm
$ rpm -Uhv --test /path/to/symlink-2-1.rpm # this should fail
$ rpm -Uhv /path/to/symlink-2-1.rpm # this should work

**Comment by dmnks** [comment]
There's also another BZ report of the same (including a reproducer which is almost identical to the one above): https://bugzilla.redhat.com/show_bug.cgi?id=1835424

**Comment by Conan-Kudo** [comment]
If we can solve this somehow, can we also bring in the old `rpmlib()` things from [this patch](https://src.fedoraproject.org/rpms/rpm/blob/f20/f/rpm-4.10.90-rpmlib-filesystem-check.patch) so that dnf installroot installs for affected distributions don't fail like this?

```
Running transaction check
Error: transaction check vs depsolve:
rpmlib(X-CheckUnifiedSystemdir) is needed by setup-2.8.48-1.fc17.noarch
rpmlib(X-CheckUnifiedSystemdir) is needed by filesystem-3-2.fc17.x86_64
To diagnose the problem, try running: 'rpm -Va --nofiles --nodigest'.
You probably have corrupted RPMDB, running 'rpm --rebuilddb' might fix the issue.
```

**Comment by pmatilai** [comment]
Eek, no. That Fedora specific hack is not coming anywhere near upstream. Sorry but I don't have a whole lot of sympathy for the use-case of installing Fedora 17 by current rpm versions...

**Comment by Conan-Kudo** [comment]
But you shipped it and then unshipped it, so now you can't even do mock or chroot builds of those releases. If handling stuff like this is properly fixed in RPM, there's no reason to not have some way to emit those `rpmlib()` capabilities either.

**Comment by pmatilai** [comment]
Those releases are something like ten years out of support by now. If you need it, the patch is out there if you need it. That's open-source power for you.

Now, this issue here is about a very specific case regarding the symlink detection logic in rpm, that actually has zero to do with the old Fedora hacks. So I'm going to hide this side-discussion as off-topic, which it is.

**Comment by seanyoung** [comment]
I am trying to use the `%pretrans` workaround and it does not work for downgrade.

In lirc-0.10.0 `/usr/include/lirc/media` is a symlink. In lirc-0.10.2 it is a directory - makes much more sense.

I've created a PR to upgrade lirc from 0.10.0 to 0.10.2 with the `%pretrans` script to work around the symlink -> directory problem, see https://src.fedoraproject.org/rpms/lirc/pull-request/6

However, the CI tests downgrade from 0.10.2 to 0.10.0 which fails because now we are replacing a directory with a symlink.

Not sure what the right solution is here.
